pkg/controller/populators/import-populator.go

```diff
@@ -0,0 +1,304 @@
+	reconciler := &ImportPopulatorReconciler{
+		ReconcilerBase: ReconcilerBase{
+			client:          client,
+			scheme:          mgr.GetScheme(),
+			log:             log.WithName(importPopulatorName),
+			recorder:        mgr.GetEventRecorderFor(importPopulatorName),
+			featureGates:    featuregates.NewFeatureGates(client),
+			sourceKind:      cdiv1.ImportSourceRef,
+			installerLabels: installerLabels,
+		},
+	}
+
+	importPopulator, err := controller.New(importPopulatorName, mgr, controller.Options{
+		Reconciler: reconciler,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if err := addCommonPopulatorsWatches(mgr, importPopulator, log, cdiv1.ImportSourceRef, &cdiv1.ImportSource{}); err != nil {
+		return nil, err
+	}
+
+	return importPopulator, nil
+}
+
+// Reconcile the reconcile loop for the PVC with DataSourceRef of ImportSource kind
+func (r *ImportPopulatorReconciler) Reconcile(_ context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("PVC", req.NamespacedName)
+	log.V(1).Info("reconciling Import Source PVCs")
+	return r.reconcile(req, r, log)
+}
+
+// Implementations of populatorController methods
+
+// Import-specific implementation of getPopulationSource
+func (r *ImportPopulatorReconciler) getPopulationSource(namespace, name string) (client.Object, error) {
+	importSource := &cdiv1.ImportSource{}
+	importSourceKey := types.NamespacedName{Namespace: namespace, Name: name}
+	if err := r.client.Get(context.TODO(), importSourceKey, importSource); err != nil {
+		// reconcile will be triggered once created
+		if k8serrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return importSource, nil
+}
+
+// Import-specific implementation of reconcileTargetPVC
+func (r *ImportPopulatorReconciler) reconcileTargetPVC(pvc, pvcPrime *corev1.PersistentVolumeClaim) (reconcile.Result, error) {
+	phase, _ := pvcPrime.Annotations[cc.AnnPodPhase]
+	if err := r.updateImportProgress(phase, pvc, pvcPrime); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	switch phase {
+	case string(corev1.PodRunning):
+		// We requeue to keep reporting progress
+		return reconcile.Result{RequeueAfter: 2 * time.Second}, nil
+	case string(corev1.PodFailed):
+		// We'll get called later once it succeeds
+		r.recorder.Eventf(pvcPrime, corev1.EventTypeWarning, importFailed, messageImportFailed, pvc.Name)
+	case string(corev1.PodSucceeded):
+		// Once the import is succeeded, we rebind the PV from PVC' to target PVC
+		if err := r.rebindPV(pvc, pvcPrime); err != nil {
+			return reconcile.Result{}, err
+		}
+		r.recorder.Eventf(pvc, corev1.EventTypeNormal, importSucceeded, messageImportSucceeded, pvc.Name)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// Import-specific implementation of updateAnnotations
+func (r *ImportPopulatorReconciler) updateAnnotations(pvc *corev1.PersistentVolumeClaim, source client.Object) {
+	importSource := source.(*cdiv1.ImportSource)
+	annotations := pvc.Annotations
```

Comment:
are we sure we want to copy all the target pvc annotations?

Reply:
I'll try to improve that.